src/bootstrap/src/core/build_steps/llvm.rs

```diff
@@ -849,6 +850,76 @@ fn get_var(var_base: &str, host: &str, target: &str) -> Option<OsString> {
         .or_else(|| env::var_os(var_base))
 }
 
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub struct Enzyme {
+    pub target: TargetSelection,
+}
+
+impl Step for Enzyme {
+    type Output = PathBuf;
+    const ONLY_HOSTS: bool = true;
+
+    fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
+        run.path("src/tools/enzyme/enzyme")
+    }
+
+    fn make_run(run: RunConfig<'_>) {
+        run.builder.ensure(Enzyme { target: run.target });
+    }
+
+    /// Compile Enzyme for `target`.
+    fn run(self, builder: &Builder<'_>) -> PathBuf {
+        builder.require_submodule(
+            "src/tools/enzyme",
+            Some("The Enzyme sources are required for autodiff."),
+        );
+        if builder.config.dry_run() {
+            let out_dir = builder.enzyme_out(self.target);
+            return out_dir;
+        }
+        let target = self.target;
+
+        let LlvmResult { llvm_config, .. } = builder.ensure(Llvm { target: self.target });
+
+        let out_dir = builder.enzyme_out(target);
+        let done_stamp = out_dir.join("enzyme-finished-building");
+        if done_stamp.exists() {
+            return out_dir;
+        }
```

Comment:
Question: would stale artifacts be a problem for non-fresh checkouts that have previously built the enzyme tool, but then say enzyme has an update and I try to build again based on the previously built artifacts?

Reply:
Good catch. We used to have a better logic here, I apparently messed that up while reimplementing, see this link.
https://github.com/EnzymeAD/rust/blame/d5d720468b5366ce101a1f20aa5ade949b6f4124/src/bootstrap/src/core/build_steps/llvm.rs#L847

 With the latest commit that I just pushed it now again handles submodule updates correctly.

Reply:
Yes, that looks good to me.

Remark (not for this PR): we should extract the submodule stamping logic to its own helper to avoid having to reinvent/copy-pasta this for every submodule and to help avoid missing stamping logic.